system/security/LdapSecurity/ldapsecurity.cpp

```diff
@@ -620,16 +620,14 @@ void CLdapSecManager::init(const char *serviceName, IPropertyTree* cfg)
 
     m_ldap_client.setown(ldap_client);
     m_pp.setown(pp);
-    int cacheTimeoutMinutes = cfg->getPropInt("@cacheTimeout", DEFAULT_RESOURCE_CACHE_TIMEOUT_MINUTES);//config value is in minutes
+    unsigned cacheTimeoutMinutes = cfg->getPropInt("@cacheTimeout", DEFAULT_RESOURCE_CACHE_TIMEOUT_MINUTES);//config value is in minutes
 
     if (cfg->getPropBool("@sharedCache", true))
-        m_permissionsCache.setown(CPermissionsCache::getInstance(cfg->queryProp("@name")));
+        m_permissionsCache.setown(CPermissionsCache::getInstance(cfg->queryProp("@name"), this, cacheTimeoutMinutes));
```

Comment:
What is this getInstance code doing?  It looks highly dangerous - is something preventing a cache object being used by two ldap managers at the same time?  When does that cache actually come into play.

This is important because your change has altered the semantics.  Previously the ldap manager was always updated in the cache object - now it is only on construction.  If this cache is ever used, the second use will lead it pointing to a released ldap manager and it will crash.

Reply:
The relationship between CPermissionsCache and CLdapSecManager objects makes no sense.
It is a pre-existing problems that if CPermissionsCache's are ever shared (the default if secMgrClass matches). The last call to secMgrClass::getInstance set m_secMgr, and there's nothing to stop that sec. manager, being destroyed whilst others use the cache.

I suspect there should always be a 1:1 relationship between a CLdapSecManager and a CPermissionsCache. It may make sense to have multiple CLdapSecManager's, but probably only 1 of each type (each with their own CPermissionsCache). 

If we can ensure that it doesn't pre-load the filescope cache per CPermissionsCache instance now, can get rid of the @shared mechanism altogether and assert there's a 1:1 relationship between these two object?